File: mobile/library/kotlin/io/envoyproxy/envoymobile/stats/DistributionImpl.kt
package io.envoyproxy.envoymobile

import io.envoyproxy.envoymobile.engine.EnvoyEngine
import java.lang.ref.WeakReference

/**
 * Envoy implementation of a `Distribution` for measurements of quantile data for int values
 */
internal class DistributionImpl : Distribution {
  var envoyEngine: WeakReference<EnvoyEngine>
  var series: String
  var tags: Tags

  internal constructor(engine: EnvoyEngine, elements: List<Element>, tags: Tags = TagsBuilder().build()) {
    this.envoyEngine = WeakReference<EnvoyEngine>(engine)
    this.series = elements.joinToString(separator = ".") { it.value }
    this.tags = tags
  }

  override fun recordValue(value: Int) {
    envoyEngine.get()?.recordHistogramValue(series, this.tags.allTags(), value)
  }

  override fun recordValue(tags: Tags, value: Int) {
    envoyEngine.get()?.recordHistogramValue(series, tags.allTags(), value)
  }
}

File: mobile/library/swift/stats/DistributionImpl.swift
@_implementationOnly import EnvoyEngine
import Foundation

/// The implementation of distribution tracking quantile/sum/average stats
@objcMembers
final class DistributionImpl: NSObject, Distribution {
  private let series: String
  private let tags: Tags
  private weak var engine: EnvoyEngine?

  init(elements: [Element], tags: Tags, engine: EnvoyEngine) {
    self.series = elements.map { $0.value }.joined(separator: ".")
    self.engine = engine
    self.tags = tags
    super.init()
  }

  /// Record a new int value for the distribution.
  func recordValue(value: Int) {
    // TODO(jingwei99) potentially surface error up if engine is nil.
    self.engine?.recordHistogramValue(
      self.series, tags: self.tags.allTags(), value: numericCast(value))
  }

  /// Record a new int value for the distribution with tags.
  func recordValue(tags: Tags, value: Int) {
    // TODO(jingwei99) potentially surface error up if engine is nil.
    self.engine?.recordHistogramValue(self.series, tags: tags.allTags(), value: numericCast(value))
  }
}
